### PR TITLE
WT-11613 Extend WT functional model to support prepared transactions

### PR DESCRIPTION
**Pull request by pmacko86**

### CODE REVIEW

File: test/model/src/core/kv_table.cpp
Comment (korteland) [inline test/model/src/core/kv_table.cpp:69]: Rather than having each function that could return an error construct an `ostringstream`, build a string, and raise an exception - what about a helper method that just takes `ret` and throws an exception? You'd lose the function name (without some macro hackery), but the traceback will give you that anyway.

File: test/model/src/include/model/kv_table.h
Comment (korteland) [inline test/model/src/include/model/kv_table.h:136]: This API is a little confusing - we now take a timestamp and a durable timestamp. It's possible to puzzle out that `timestamp` must be the commit timestamp since the comments says it's commit and durable, and the other is `durable_timestamp`. But that's a lot of work, and easy to get wrong at the callsite.

I would suggest either:
- Having different `fix_commit_timestamp` and `fix_durable_timestamp` functions
- A `fix_timestamp` function that takes a `timestamp_t` and an enum describing which timestamp to set

File: test/model/src/include/model/kv_table_item.h
Comment (korteland) [inline test/model/src/include/model/kv_table_item.h:77]: `const`? Here and a bunch of other places in this class.

File: test/model/src/core/kv_table_item.cpp
Comment (korteland) [inline test/model/src/core/kv_table_item.cpp:227]: I feel like this was discussed in the previous PR, but the proliferation of out parameters here makes me think `std::optional` is worth considering.

File: test/model/src/core/kv_transaction.cpp
Comment (korteland) [inline test/model/src/core/kv_transaction.cpp:181]: Do you think it'd be worth factoring these out into something like `assert_valid_transition(state old, state new)`? It gets a little repetitive having to think about all possible combinations in several functions.

File: test/model/src/include/model/kv_table_item.h
Comment (pmacko86) [reply test/model/src/include/model/kv_table_item.h:77]: I wish this would be possible, but these functions cannot be made `const` because they take lock (or call functions that take lock), so they actually need read-write access to the instance's fields.

File: test/model/src/core/kv_table_item.cpp
Comment (pmacko86) [reply test/model/src/core/kv_table_item.cpp:227]: I'm not sure how this would help. Can you please elaborate? `data_value` already has a `NONE` value, which was returned for when the value was deleted, but I needed to make this change as reading can now return WiredTiger errors.

If we don't like the `out` argument, I can think of two options:
* Return an instance of `std::variant<int, model::data_value>`, which can be used to represent either an error or a valid result. I decided against it, as returning values via parameters is a lot more common in C/C++; I've only really seen this alternative in Go, and I don't think it's particularly common in our code, so I didn't want to unnecessarily introduce a new non-standard way of doing it.
* Throw an exception, so that we can continue to return a `data_value`. This would be more C++ like, but I ended up not doing this, so that the top-level function can return a WiredTiger compatible return code.

If there is desire to remove the `out` arguments, I can switch to the latter approach with exceptions, and only translate them to WiredTiger-compatible int return codes at the top-level `kv_table` API, so that they can be compared directly to what's returned from regular WiredTiger function calls.

File: test/model/src/core/kv_transaction.cpp
Comment (pmacko86) [reply test/model/src/core/kv_transaction.cpp:181]: I refactored this check into `assert_in_progress_or_prepared()`, as this is the most common case. I don't think refactoring state transition checks outside the relevant function would make sense so that we can, for example, keep all commit logic together.

File: test/model/src/core/kv_table_item.cpp
Comment (pmacko86) [reply test/model/src/core/kv_table_item.cpp]: After thinking about this more, I ended up switching to throwing exceptions on error, so that we can keep returning `data_value`.

File: test/model/src/core/kv_table.cpp
Comment (pmacko86) [reply test/model/src/core/kv_table.cpp]: This is now moot—I'm throwing a `wiredtiger_exception` in the upstream code on error.

**Comment by pmacko86** [comment]
For consistency, I refactored the internal error handling to use exceptions instead of return codes, and only translate exceptions for return codes at the top-level API, e.g., `kv_table::insert()`. This should provide clearer expectations within the model's code instead of mixing error handling paradigms.

I'm actually even just wondering if it would make sense to go all the way and use only exceptions within the model code, even at the top-level API, and just modify tests to handle those exceptions instead of expecting return codes from functions such as `insert` or `update`.

File: test/model/src/include/model/kv_table.h
Comment (korteland) [reply test/model/src/include/model/kv_table.h:136]: OK, that makes more sense now. Maybe just rename the `timestamp` parameter to `commit_timestamp` then?

File: test/model/src/include/model/kv_table_item.h
Comment (korteland) [reply test/model/src/include/model/kv_table_item.h:77]: Ah, that's annoying. I guess if you were really keen you could mark the mutex as `mutable` and then get away with these functions being `const` - but that's optional and there's no need to block the PR on it.

File: test/model/src/core/kv_table_item.cpp
Comment (korteland) [reply test/model/src/core/kv_table_item.cpp]: Thanks - the code reads better now, at least to me.

File: test/model/src/core/kv_transaction.cpp
Comment (korteland) [reply test/model/src/core/kv_transaction.cpp]: Fair enough - thanks.

File: test/model/src/include/model/kv_table.h
Comment (pmacko86) [reply test/model/src/include/model/kv_table.h:136]: Ah, that's right. I thought I renamed it, but I renamed it in only one place and not here. Ok, done. Thanks for catching that!

File: test/model/src/include/model/kv_table_item.h
Comment (pmacko86) [reply test/model/src/include/model/kv_table_item.h:77]: I didn't know about this keyword! Let me use it—hopefully it helps.

**Review by korteland — APPROVED** [review]
Thanks for the tidyups - LGTM!

File: test/model/src/core/kv_transaction.cpp
Comment (ddanderson) [inline test/model/src/core/kv_transaction.cpp:124]: Comment needs fixing.

**Review by ddanderson — APPROVED** [review]
LGTM!

**Pull request merged**
